store/tikv/region_cache.go

```diff
@@ -939,15 +960,25 @@ func (c *RegionCache) switchToPeer(r *Region, targetStoreID uint64) (found bool)
 	return
 }
 
-func (c *RegionCache) switchNextPeer(r *Region, currentPeerIdx int) {
-	regionStore := r.getStore()
-	if int(regionStore.workStoreIdx) != currentPeerIdx {
+func (c *RegionCache) switchNextPeer(r *Region, currentPeerIdx int, err error) {
+	rs := r.getStore()
+	if int(rs.workStoreIdx) != currentPeerIdx {
 		return
 	}
-	nextIdx := (currentPeerIdx + 1) % len(regionStore.stores)
-	newRegionStore := regionStore.clone()
+
+	if err != nil { // TODO: refine err, only do this for some errors.
+		s := rs.stores[rs.workStoreIdx]
+		epoch := rs.storeFails[rs.workStoreIdx]
```

Comment:
It should be `atomic.load`? Or maybe a lock will be better. I'm not sure are there any other threads can access `storeFails` concurrently.

Reply:
good catch, this no need `atomic.load`, because `r.getStore` does that, the whole `rs` need follow copy-on-write idiom

so https://github.com/pingcap/tidb/pull/11344/files#diff-708f6242b27e2b7bcf0e905e9b0eacf2R973 has bug, and we should not +1 for `rs.storeFails[rs.workStoreIdx]` at here, because after try other peer, region maybe back to current idx, and we need reload region at that time.